acme: added option `security.acme.preliminarySelfsigned`

bobvanderlinden opened the pull request:
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [x] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
- [ ] Documentation
- [ ] Code cleanup

---

Consider this WIP. This adds better Nginx integration for ACME. When Nginx was used with ACME, there's a chicken+egg problem: ACME needs Nginx to host files over HTTP, the Nginx configuration needs certificates to actually start. This problem (and solution) was covered in #12655. This PR makes sure Nginx always has certificates: if they are non-existant, it'll create self-signed ones. Once ACME created the proper certificates, Nginx is reloaded. This is based on the solution from @domenkozar.

Basically it adds the option `security.acme.enableNginxIntegration`. When enabled, it creates an one-shot service for each ACME cert that runs before Nginx. It creates a self-signed certificate if there is no certificate yet.

I'd like to know whether this PR is the right direction. Maybe we want a single service that creates self-signed certificates for all ACME certs, instead of separate services? Separate services have the advantage that they can be individually disabled. A single service will probably make the Nix expression less complex.

I also considered doing this in a separate module, but since I wanted `security.acme.enableNginxIntegration` I needed to add the option to acme.nix anyway. Let me know what is considered best-practice in this regard.

For apache/httpd support the same service can be reused. I haven't added this in this PR because I don't have a apache setup atm.

bobvanderlinden committed:
acme: added option `security.acme.preliminarySelfsigned`

groxxda commented:
Can we try to remove the nginx specific parts on this?
I'd rename `enableNginxIntegration` to something like `preliminarySelfsigned`.
Remove the reload nginx command from PostStop and. In my opinion the nginx service should watch for file changes of the certificate and execute reload accordingly, this way it will also automatically pickup a change when the timer renews the certificate.

Will the initial ACME certificate-request be triggered automatically when nginx is started? Or will it wait until the timer is fired the first time?

The dependency could be made configurable so a user can pass in "nginx.service" or "httpd.service" which is then used for the fake-cert-service.(WantedBy,Before).

groxxda commented:
CC @globin (module maintainer)

bobvanderlinden commented:
@groxxda Indeed, having it independent of nginx would be nice. I've changed the PR to have `preliminarySelfsigned`. When enabled, the `acme-selfsigned-certificates` and `acme-certificates` targets can be used from any other service to specify when it needs to start.

Example configuration:

```
  security.acme.preliminarySelfsigned = true;
  systemd.targets."acme-selfsigned-certificates" = {
    before = [ "nginx.service" ];
    wantedBy = [ "nginx.service" ];
  };
  systemd.targets."acme-certificates" = {
    wantedBy = [ "nginx.service" ];
  };
```

groxxda commented on nixos/modules/security/acme.nix:
```diff
@@ -140,54 +145,114 @@ in
+                  wantedBy = [ "acme-certificates.target" ];
+                };
+                selfsignedService = {
+                  description = "Create self-signed ACME Certificate for ${cert}";
```
Create preliminary self-signed certificate for ${cert}

groxxda commented on nixos/modules/security/acme.nix:
```diff
@@ -140,54 +145,114 @@ in
+                      if [ ! -d '${data.webroot}' ]; then
+                        mkdir -p '${data.webroot}'
+                        chmod -R ${rights} '${data.webroot}'
+                        chown -R 'root:root' '${data.webroot}'
```
should probably match the user:group of the real acme certificate

groxxda commented on nixos/modules/security/acme.nix:
```diff
@@ -140,54 +145,114 @@ in
+                      then
+                        mkdir -p '${cpath}'
+                        chmod ${rights} '${cpath}'
+                        chown 'root:root' '${cpath}'
```
same here

groxxda commented on nixos/modules/security/acme.nix:
```diff
@@ -140,54 +145,114 @@ in
+                        rmdir $workdir
+
+                        chmod ${rights} '${cpath}/key.pem'
+                        chown 'root:root' '${cpath}/key.pem'
```
same here

groxxda commented on nixos/modules/security/acme.nix:
```diff
@@ -140,54 +145,114 @@ in
+                        chmod ${rights} '${cpath}/key.pem'
+                        chown 'root:root' '${cpath}/key.pem'
+                        chmod ${rights} '${cpath}/fullchain.pem'
+                        chown 'root:root' '${cpath}/fullchain.pem'
```
and here

groxxda commented on nixos/modules/security/acme.nix:
```diff
@@ -140,54 +145,114 @@ in
+                  description = "Create self-signed ACME Certificate for ${cert}";
+                  script = 
+                    ''
+                      if [ ! -d '${data.webroot}' ]; then
```
`data.webroot` is not needed for the preliminary certificate

groxxda commented on nixos/modules/security/acme.nix:
```diff
@@ -140,54 +145,114 @@ in
+                        chmod ${rights} '${cpath}'
+                        chown 'root:root' '${cpath}'
+                      fi
+                      if [ ! -f '${cpath}/key.pem' ]; then
```
could be converted to the systemd.unit option `ConditionPathExists=!${cpath}/key"` to bail out earlier

groxxda commented on nixos/modules/security/acme.nix:
```diff
@@ -140,54 +145,114 @@ in
+                      fi
+                      if [ ! -f '${cpath}/key.pem' ]; then
+                        # Create self-signed key
+                        workdir="$(mktemp -d)"
```
Just an idea: use `systemd.exec.RuntimeDirectory` and let systemd handle the creation and cleanup of the temporary files. I'm not sure if this works for oneshot though.

groxxda commented:
Thanks for making it independent so quickly :+1: 
I added a few more comments. Maybe it's also a good idea to create the output directory in preStart and then set `systemd.service.PermissionsStartOnly` and `systemd.exec.{User,Group}` just like the acme service does. This way openssl client will not be run as root.

I'm not sure if it's worth it though.

cc @fpletz

bobvanderlinden commented:
@groxxda I've integrated your suggestions including using `PermissionsStartOnly`. It seemed fitting as the original ACME services uses that as well.

The documentation now includes examples on how to use `security.acme.preliminarySelfsigned` with Nginx.

I still would like there to be some kind of simple way to make Nginx (or Apache) work without needing to copy example-configuration from the documentation though. This would require reloading Nginx when the files are changing (even though Nginx doesn't do this out-of-the-box). It might be a good idea to create systemd.path files that watch the certificates and reload Nginx, but that would be something that needs to be configured in the Nginx module. Probably outside the scope of this PR atm.

groxxda commented on nixos/modules/security/acme.xml:
```diff
@@ -66,4 +66,31 @@ options for the <literal>security.acme</literal> module.</para>
 </section>
 
+<section><title>Using ACME certificates in Nginx</title>
+<para>In practice ACME is mostly used for retrieval &amp; of certificates that will be used in a webserver like Nginx. A configuration for Nginx that uses the certificates from ACME for <literal>foo.example.com</literal> will look similar to:</para>
```
`retrieval &` ? renewal?

bobvanderlinden replied:
Thanks, missed that one. Fixed in PR.

groxxda commented:
Looks mergeable to me :tada: for the documentation.

Out of the original scope of this PR but might be added: acme service will not work without routable network so it should really have `After = network-online.target` and `Requires = network-online.target`.

Using systemd.path is a good idea but should be added to nginx module itself and is out of scope of this PR

domenkozar commented on nixos/modules/security/acme.xml:
```diff
@@ -66,4 +66,31 @@ options for the <literal>security.acme</literal> module.</para>
+<programlisting>
+security.acme.preliminarySelfsigned = true;
+systemd.services.nginx = {
+  after = [ "acme-selfsigned-certificates.target" ];
```
Couldn't we put this into nginx itself? I think if target is missing in general, systemd will just skip it.

bobvanderlinden replied:
Yes, that's a possibility, but I can imagine that ACME could be for Httpd,
while Nginx is used for something else. In that case Nginx doesn't need to
be depending on `acme-selfsigned-certificates.target`.

That said, if the intention is to de-clutter the machine-configuration with
ACME+nginx specific workarounds, I agree. We might need to improve upon
this, but this PR already brings us a step further.

I'm still thinking of having explicit integration options.
`security.acme-integration.nginx.enable` or something similar. It could be
a separate module on top of ACME that makes use of the
`security.acme.preliminarySelfsigned` option. That way each webserver-acme
integration can implement their own additional options that are needed. For
nginx you would need to depend on `acme-selfsigned-certificates.target` and
watch for the certificate files to reload nginx. apache might already
reload when certificate files change, but for apache you could
automatically create the certificate definitions from
`services.httpd.virtualHosts.*.hostName`
and `services.httpd.virtualHosts.*.enableSSL` or maybe even automatic http
-> https redirects. It seems obvious to me to have such options when HTTPS
is becoming the default and ACME is becoming the default for HTTPS. At the
moment to use ACME correctly someone would need to know quite a bit about
how services are related to eachother.

On Mon, May 23, 2016 at 2:00 PM Domen Kožar <notifications@github.com>
wrote:

> In nixos/modules/security/acme.xml
> <https://github.com/NixOS/nixpkgs/pull/15562#discussion_r64210274>:
>
> > +    listen 443 ssl;
> > +    ssl_certificate     ${config.security.acme.directory}/foo.example.com/fullchain.pem;
> > +    ssl_certificate_key ${config.security.acme.directory}/foo.example.com/key.pem;
> > +    root /var/www/foo.example.com/;
> > +  }
> > +'';
> > +</programlisting>
> > +
> > +<para>Now Nginx will try to use the certificates that will be retrieved by ACME. However, Nginx will not start until the certificates actually exist. Additionally, ACME will not be able to retrieve certificates until Nginx can host the challenge-files for ACME over HTTP. This results in a problem where the two services need eachother to function.</para>
> > +
> > +<para>To circumvent this problem we can generate self-signed certificates that Nginx can use until ACME has retrieved new ones. This can be done manually, but the ACME module can also do this automatically with the <literal>security.acme.preliminarySelfsigned</literal> option. If we use this option we need to make sure Nginx is only started after the self-signed certificate is generated. This can be configured using:</para>
> > +
> > +<programlisting>
> > +security.acme.preliminarySelfsigned = true;
> > +systemd.services.nginx = {
> > +  after = [ "acme-selfsigned-certificates.target" ];
>
> Couldn't we put this into nginx itself? I think if target is missing in
> general, systemd will just skip it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/15562/files/350ee207b8c1ab2bf5ce2d4af709dc9398e5af3a#r64210274>
>

bobvanderlinden replied:
@domenkozar Should the target be added to nginx in this PR?

domenkozar replied:
For now, we could add the nginx target declaration from acme module if nginx.enable is true. Then devs can go into micro optimizations of they run apache and nginx (1% case I wouldn't care about now).

bobvanderlinden commented:
Alright, added nginx rules to make sure it is started after self-signed certificates are created. Removed the section from the docs where it instructs people to add this themselves.

Profpatsch commented:
Do we need an explicit configuration option? If for example nginx & acme are enabled, does the user have any reason to disable the bootstrap process? I can’t think of any.

Since you generalized it, maybe configure a few combinations that use it by default:

```nix
config = mkIf c.security.acme.enabled {
  security.acme.preliminarySelfSigned = 
    mkIf (c.services.nginx.enabled || c.services.httpd.enabled) true;
}
```

bobvanderlinden commented:
@Profpatsch If it should be enabled by default, I'd rather go for `default = true` in `mkOption`. Enabling it based on `nginx.enable` and `httpd.enable` adds more complexity than it's worth. Having it set in `mkOption` is more obvious for people that look at the option documentation that way. Otherwise it says `default = false`, but is actually enabled in most cases.

I do agree that having this enabled by default would help out most people. There seem to be little to no negative consequences to creating the self-signed certificates first.

Let me know if anyone is against enabling this by default.

Profpatsch commented:
You are right. I haven’t reviewed the code, but +1 if it works. :P

bobvanderlinden commented:
Ok I think this is ready to be merged. To summarize, this PR:
* Creates self-signed certificates before creating actual ACME certificates
* Makes sure nginx is started after the self-signed certificates are created
* Adds option `security.acme.preliminarySelfsigned` to turn off creating self-signed certificates

Profpatsch commented:
ping @fpletz

bobvanderlinden commented:
I've added some more explanation to the .nix file and squashed the commits into one for merging.

I've tested whether nginx starts when it's configured to use acme using: https://gist.github.com/bobvanderlinden/6c604ccb3f5a0127a117f269a674ae7a. It unfortunately cannot test for real ACME support.
My development server now has a similar addition. It seems to work nicely there.

edolstra commented on nixos/modules/security/acme.xml:
```diff
@@ -66,4 +66,23 @@ options for the <literal>security.acme</literal> module.</para>
 </section>
 
+<section><title>Using ACME certificates in Nginx</title>
+<para>In practice ACME is mostly used for retrieval &amp; renewal of certificates that will be used in a webserver like Nginx. A configuration for Nginx that uses the certificates from ACME for <literal>foo.example.com</literal> will look similar to:</para>
```
`&amp;` -> and.

edolstra replied:
Please break lines at ~80 characters, otherwise it's hard to read e.g. in Github.

bobvanderlinden replied:
Thanks, these are fixed now. `&amp;` is still being used else where in `acme.xml` though. I'll leave those for a separate PR.

edolstra commented on nixos/modules/security/acme.xml:
```diff
@@ -66,4 +66,23 @@ options for the <literal>security.acme</literal> module.</para>
+
+<para>Now Nginx will try to use the certificates that will be retrieved by ACME. However, Nginx will not start until the certificates actually exist. Additionally, ACME will not be able to retrieve certificates until Nginx can host the challenge-files for ACME over HTTP. This results in a problem where the two services need eachother to function.</para>
+
+<para>To circumvent this problem we can generate self-signed certificates that Nginx can use until ACME has retrieved new ones. This can be done manually, but the ACME module can also do this automatically with the <literal>security.acme.preliminarySelfsigned</literal> option. The ACME module contains the necessairy rules for Nginx to start after the self-signed certificates are created.</para>
```
"necessairy"

domenkozar commented:
I've deployed this on `16.03` with following example:

```
  services.nginx.httpConfig = ''
    server {
      server_name          kamdanes.si;
      listen               80;

      location /.well-known/acme-challenge {
        root /var/www/challenges;
      }

      location / {
        return 301 https://$host$request_uri;
      }

    }

    server {
      server_name kamdanes.si;
      listen 443 ssl;
 
      ssl_certificate     ${config.security.acme.directory}/kamdanes.si/fullchain.pem;
      ssl_certificate_key ${config.security.acme.directory}/kamdanes.si/key.pem;

      location / {
        root ${kamdanesFrontend};
        index index.html;
      }

      location /api/ {
        proxy_pass http://localhost:3000/;
        proxy_set_header Host $http_host;
        proxy_set_header REMOTE_ADDR $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      }
    }
  '';

  security.acme.certs."kamdanes.si" = {
    email = "domen@dev.si";
    user = "nginx";
    group = "nginx";
    webroot = "/var/www/challenges";
    postRun = "systemctl reload nginx.service";
  };
```

Notes:

- it would be useful for nginx integration example to be self-contained. Copy/paste, replace domains, etc and it should work
- `webroot` is not automatically mkdired, could also do that?
- since `preliminarySelfSigned` is now on by default, last part of documentation addition is confusing, I would just remove it.
- to use this on `16.03`, replace`openssl.bin` with `openssl`

Great work!!

fpletz commented:
Also looks too to me. Thank you very much! :beers: 

One thing: We've added support for acme to the nginx module which we will finally PR later today. I think the nginx service ordering hack shouldn't be in the acme module but rather dependent services like nginx should order themselves if they depend on it. We can merge it in the current state, though, and later remove it with the nginx PR.

bobvanderlinden commented:
Thanks for all the feedback.

@domenkozar The last part of the documentation is now removed/altered.
`mkdir` of webroot is somewhat outside the scope of this PR. For preliminary self-signed certificates the webroot isn't really needed. I can do a separate PR, because I think the code needs some cleanup/refactor anyway. Maybe after or combined with the PR of @fpletz?

@fpletz Sounds good. The ordering is related to both ACME and Nginx. I was thinking of having separate integration modules for nginx<->acme that can be enabled separately, but that might be overkill.
I'm interested in seeing the additional ACME - Nginx support PR!

domenkozar commented:
@bobvanderlinden sure it's out of scope, but I'd like acme example to be self contained. I'm all for adding docs it has to be `mkdir`-ed, but I'd never expect that to be a requirement.

domenkozar commented:
I see now this part is already documented at http://nixos.org/nixos/manual/index.html#module-security-acme - maybe we should just change behavior to chown to the acme user. But I agree it's out of scope.

domenkozar commented:
I'm merging this for now as it works as documented. This is very close to being smooth :D Thank you!

bobvanderlinden commented:
Thanks!
